Adding fluxcd-source-controller 0.21.2 for TCE

EstebanFS opened the pull request:
## What this PR does / why we need it
This PR adds the [fluxcd-source-controller](https://github.com/fluxcd/source-controller) package as an addon having the Package bits coming out of [vmware-tanzu/package-for-source-controller](https://github.com/vmware-tanzu/package-for-source-controller).

## Details for the Release Notes (PLEASE PROVIDE)
```release-note
Add fluxcd-source-controller v0.21.2 package
```
## Which issue(s) this PR fixes
Fixes: #3087

## Describe testing done for PR

1. Create a cluster provisioned with kapp-controller and cert-manager
2. Submit the metadata.yaml and 0.2.2/package.yaml files
3. Install the package using the tanzu cli

```text
tanzu package install fluxcd-source-controller -p fluxcd.source.controller.community.tanzu.vmware.com -v 0.21.2
\ Installing package 'fluxcd.source.controller.community.tanzu.vmware.com'
| Getting package metadata for 'fluxcd.source.controller.community.tanzu.vmware.com'
| Creating service account 'fluxcd-source-controller-default-sa'
| Creating cluster admin role 'fluxcd-source-controller-default-cluster-role'
| Creating cluster role binding 'fluxcd-source-controller-default-cluster-rolebinding'
| Creating package resource
/ Waiting for 'PackageInstall' reconciliation for 'fluxcd-source-controller'
\ 'PackageInstall' resource install status: Reconciling


 Added installed package 'fluxcd-source-controller'
```
run the smoke test from package-for-source-controller (see https://github.com/vmware-tanzu/package-for-source-controller/blob/main/test/e2e.sh)

you can also see more deeper tests executed on our pipeline:

- [test-installation](https://runway-ci.eng.vmware.com/teams/tce-fluxcd-source-controller/pipelines/tce-flux-source-controller/jobs/test/builds/19#L621c6b83:31:43)
- [test -smoke](https://runway-ci.eng.vmware.com/teams/tce-fluxcd-source-controller/pipelines/tce-flux-source-controller/jobs/test/builds/19#L621c6b84:35:143)

EstebanFS committed:
Adding fluxcd-source-controller 0.21.2 for TCE

jorgemoralespou commented on addons/packages/fluxcd-source-controller/0.21.2/package.yaml:
```diff
@@ -0,0 +1,51 @@
+                    spec:
+                      #@overlay/match missing_ok=True
+                      imagePullSecrets:
+                      - name: canonical-registry-secret
```
I can't find any documentation on what this secret is for and how it's added to the cluster. I understand that secretgen injects the canonical-secret-above with .... for .... is what should be documented.

EstebanFS replied:
You are right!, there is not use for that secret since all the images the project is going to use are public. That conf come from TAP because some of the images where hosted in private repos!

Let me do a quick fix for that!

jorgemoralespou commented:
I'm missing in the documentation some description of what's generated. (e.g. a namespace with fluxcd is created, this namespace is fixed and can not be changed. Once the package is installed the user can verify fluxcd is working by looking at whatever pods, or logs,....)

Also, I'm missing what the canonical-registry-secret is for. A user most likely will not know how to populate that secret, as secretgen is a hidden feature not properly documented in TCE.

Finally, looking at the sources of the provided package `projects.registry.vmware.com/tce/fluxcd-source-controller-bundle@sha256:020f9efbc07d439bc9bc1857abdca053a9470eb98ba5d5af8de1fa43db141c93` It seems that there's one configruable value that really does nothing `service_port` and that the config file is made a ytt file even if not needed. Not really an issue to this PR, but the package can probably be improved.

EstebanFS commented:
I will add some extra information on the documentation about what is generated by the package and how they could check it!.

Also will delete the service_port value since it is not doing anything there.

joshrosso commented:
Added @seemiller as a reviewer.

seemiller commented on addons/repos/main.yaml:
```diff
@@ -1,5 +1,8 @@
 ---
 packages:
+  - name: fluxcd-source-controller
+    versions:
+      - 0.21.2
   - name: cert-manager
     versions:
       - 1.3.3
```
@EstebanFS Actually, don't worry about this, since FluxCD is slated for v0.11.0, let's keep this change here so that when it's merged it will be in main and ready for the release.

seemiller reviewed (changes requested):
Would you mind adding a Usage Example?

It doesn't have to be much, just a quick, ok, now that it's installed, you can exercise this cool Foo feature by pushing Bar. Certainly not looking for a recreation of the docs or a tutorial, just a basic example.

This helps a user verify that the package is not only installed, but working as intended

EstebanFS commented:
Hey @jorgemoralespou I just updated the documentation to include a section about how to test flux, also deleted the canonical-registry-secret from the package.

There is still missing to delete the service-port from the package, I will update the pipeline so it doesn't include that

chaudharyalka commented on addons/packages/fluxcd-source-controller/0.21.2/package.yaml:
```diff
@@ -0,0 +1,51 @@
+apiVersion: data.packaging.carvel.dev/v1alpha1
+kind: Package
+metadata:
+  name: fluxcd.source.controller.community.tanzu.vmware.com.0.21.2
```
Should we keep the package name as source-controller.fluxcd.tanzu.vmware.com ?

EstebanFS replied:
should we avoid the community on the name?

I can see why swap source-controller and fluxcd, I was just following the naming giving inside TAP
[TAP-source-controller](https://gitlab.eng.vmware.com/tap/tap-packages/-/blob/master/packages/fluxcd-source-controller/0.16.3/package.yaml#L6)

to keep some kind of consistency, also I added the community word just to keep align with other packages
[kpack](https://github.com/vmware-tanzu/package-for-kpack/blob/main/metadata.yaml#L7)

EstebanFS replied:
so `fluxcd-source-controller.community.tanzu.vmware.com.0.21.2` instead of `fluxcd.source.controller.community.tanzu.vmware.com.0.21.2`

chaudharyalka commented on addons/packages/fluxcd-source-controller/metadata.yaml:
```diff
@@ -0,0 +1,14 @@
+spec:
+  displayName: "Flux Source Controller"
+  longDescription: "The source-controller is a Kubernetes operator, specialised in artifacts acquisition from external sources such as Git, Helm repositories and S3 buckets."
+  shortDescription: "The source-controller is a Kubernetes operator, specialised in artifacts acquisition from external sources such as Git, Helm repositories and S3 buckets."
```
Long description and short description is same here. Can you update the short description?

chaudharyalka commented:
I can see you are following the older approach where packages are hosted in the TCE repository under [addons/packages].
Can you please justify why you are not creating a new repo for the same?

https://github.com/vmware-tanzu/community-edition/blob/main/docs/packaging/considerations/git.md#package-in-the-tce-repository

Build-integration-team will introduce helm and kustomize controller package in TCE. We are trying to maintain consistency across these three packages structurally. We have created a temporary repo for the above two packages can we follow the same directory structure for source controller also.

https://gitlab.eng.vmware.com/calka/flux-controllers/-/tree/main/tce-packages

├── 1.0.0
│   ├── README.md
│   ├── bundle
│   │   ├── .imgpkg
│   │   │   └── images.yml
│   │   ├── config
│   │   │   ├── schema.yml
│   │   │   ├── overlays
│   │   │   │   └── overlay-a.yaml
│   │   │   ├── upstream
│   │   │   │   └── upstream-a.yaml
│   │   │   └── values.yaml
│   │   ├── vendir.lock.yml
│   │   └── vendir.yml
│   ├── package.yaml
│   └── test
│       ├── Makefile
│       ├── README.md
│       ├── e2e
│       │   └── test.go
│       └── unittest
│           └── test.go
├── LICENSE
└── metadata.yaml

chaudharyalka commented:
We need proxy support on GitRepository for which there will be a requirement to set environment variables HTTPS_PROXY and NO_PROXY in the source controller pod. 
Can you add these two env variables in the package configuration?

chaudharyalka commented on addons/packages/fluxcd-source-controller/metadata.yaml:
```diff
@@ -0,0 +1,14 @@
+  supportDescription: "https://tanzu.vmware.com/support"
+  providerName: "VMware"
+  categories:
+  - policy
```
should we update the category here? for the other fluxcd packages we are using gitops.

kartiklunkad26 replied:
Is this as simple as switching from policy to gitops in the categories section?

EstebanFS replied:
Changed!

kartiklunkad26 commented:
> We need proxy support on GitRepository for which there will be a requirement to set environment variables HTTPS_PROXY and NO_PROXY in the source controller pod. Can you add these two env variables in the package configuration?

@chaudharyalka Hey Alka - we are working on a tight timeline to release this in TCE v0.11. Could we mark this as a separate issue on this Package? cc: @EstebanFS

seemiller commented on addons/packages/fluxcd-source-controller/0.21.2/package.yaml:
```diff
@@ -0,0 +1,27 @@
+apiVersion: data.packaging.carvel.dev/v1alpha1
+kind: Package
+metadata:
+  name: fluxcd.source.controller.community.tanzu.vmware.com.0.21.2
```
```suggestion
  name: fluxcd-source-controller.community.tanzu.vmware.com.0.21.2
```

There are numerous places where this name would have to change.

EstebanFS replied:
For sure, let me change it!

EstebanFS replied:
I just changed the name of the package!

seemiller commented on addons/repos/main.yaml:
```diff
@@ -1,5 +1,8 @@
 ---
 packages:
+  - name: fluxcd-source-controller
+    versions:
+      - 0.21.2
```
@EstebanFS Can you alphabetize this? Doing that will also clear the merge conflict.

EstebanFS replied:
I just putted fluxcd-source-controller in the right spot!

seemiller reviewed (changes requested):
@EstebanFS I was reviewing the latest updates and noticed that all of the changes have been made here in the TCE source repo. All of these changes also need to be made in the [vmware-tanzu/package-for-source-controller](https://github.com/vmware-tanzu/package-for-source-controller) source repo as well.

That made me realize that we're not using vendir to sync the contents of the directories. If we use vendir, then these changes only need to be made in 1 place, the package source repository. You can then create a tag/release in the package-for-source-controller repo and we can sync the needed files over.

For example, if you make the these changes in the package-for-source-controller repo, we can then create a tag and a release. Then in the TCE `addons/packages/fluxcd-source-controller` directory we can add this vendir.yaml file and run `vendir sync` That will copy files from the package source repo and ensure that we have the same content in TCE as is declared in the "release" of the source controller package.

```yaml
apiVersion: vendir.k14s.io/v1alpha1
kind: Config
minimumRequiredVersion: 0.12.0
directories:
  - path: 0.21.2
    contents:
      - path: .
        githubRelease:
          slug: vmware-tanzu/package-for-source-controller
          tag: v0.21.2
          disableAutoChecksumValidation: true
          assetNames:
            - "package.yaml"
            - "README.md"
```

The `kpack` package is a great example of how this is done. I'm happy to pair up with you to make this happen.

EstebanFS commented:
For sure @seemiller, I will add the vendir conf, do you mind if for now I just sync everything manually?, I want to merge this PR soon

seemiller commented:
@EstebanFS I'm cool with getting the PR done, will loop in @kartiklunkad26 to ensure that we get a follow up with the vendir approach.
